protocol/blockrelay/blockrelay.go

```diff
@@ -0,0 +1,274 @@
+func readMsgBlock(msgChan <-chan wire.Message,
+	invsQueue *[]*wire.MsgInvRelayBlock) (msgBlock *wire.MsgBlock, shouldStop bool, err error) {
+
+	for {
+		const stallResponseTimeout = 30 * time.Second
+		select {
+		case <-time.After(stallResponseTimeout):
+			return nil, false, errors.Errorf("stalled for %s", stallResponseTimeout)
+		case msg, isOpen := <-msgChan:
+			if !isOpen {
+				return nil, true, nil
+			}
+
+			inv, ok := msg.(*wire.MsgInvRelayBlock)
+			if !ok {
+				return msg.(*wire.MsgBlock), false, nil
+			}
+
+			*invsQueue = append(*invsQueue, inv)
+		}
+	}
+}
+
+func deleteFromRequestedBlocks(blockHashes map[daghash.Hash]struct{}) {
+	for hash := range blockHashes {
+		hash := hash
+		requestedBlocks.delete(&hash)
+	}
+}
+
+func requestBlocks(netAdapater *netadapter.NetAdapter, router *netadapter.Router, peer *peerpkg.Peer, msgChan <-chan wire.Message,
+	dag *blockdag.BlockDAG, invsQueue *[]*wire.MsgInvRelayBlock, requestQueue *[]*daghash.Hash,
+	requestQueueSet map[daghash.Hash]struct{}) (shouldStop bool, err error) {
+
+	var hashesToRequest []*daghash.Hash
+	if len(*requestQueue) > wire.MsgGetRelayBlocksHashes {
+		hashesToRequest, *requestQueue = (*requestQueue)[:wire.MsgGetRelayBlocksHashes],
+			(*requestQueue)[wire.MsgGetRelayBlocksHashes:]
+	} else {
+		hashesToRequest, *requestQueue = *requestQueue, nil
+	}
+
+	pendingBlocks := map[daghash.Hash]struct{}{}
+	for _, hash := range hashesToRequest {
+		delete(requestQueueSet, *hash)
+		pendingBlocks[*hash] = struct{}{}
+		exists := requestedBlocks.addIfExists(hash)
+		if exists {
+			return false, nil
+		}
+	}
+
+	getRelayBlocksMsg := wire.NewMsgGetRelayBlocks(hashesToRequest)
+	router.WriteOutgoingMessage(getRelayBlocksMsg)
+
+	for len(pendingBlocks) > 0 {
+		msgBlock, shouldStop, err := readMsgBlock(msgChan, invsQueue)
+		if err != nil {
+			return false, err
+		}
+		if shouldStop {
+			return true, nil
+		}
+
+		block := util.NewBlock(msgBlock)
+		blockHash := block.Hash()
+		if _, ok := pendingBlocks[*blockHash]; !ok {
+			return false, errors.Errorf("got unrequested block %s", block.Hash())
+		}
+		delete(pendingBlocks, *blockHash)
+		requestedBlocks.delete(blockHash)
+
+		shouldStop, err = processAndRelayBlock(netAdapater, router, peer, dag, requestQueue, requestQueueSet, block)
+		if err != nil {
+			return false, err
+		}
+		if shouldStop {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func processAndRelayBlock(netAdapter *netadapter.NetAdapter, router *netadapter.Router, peer *peerpkg.Peer,
+	dag *blockdag.BlockDAG, requestQueue *[]*daghash.Hash, requestQueueSet map[daghash.Hash]struct{},
+	block *util.Block) (shouldStop bool, err error) {
+
+	blockHash := block.Hash()
+	isOrphan, isDelayed, err := dag.ProcessBlock(block, blockdag.BFNone)
+	if err != nil {
+		// When the error is a rule error, it means the block was simply
+		// rejected as opposed to something actually going wrong, so log
+		// it as such. Otherwise, something really did go wrong, so log
```

Comment:
You are not "logging it as an actual error" - you are killing the whole application.
Is fine, but comment should imply the gravity of the situation